services/subscriber/service.go

```diff
@@ -25,11 +27,66 @@ const (
 	statWriteFailures  = "writeFailures"
 )
 
+// WriteRequest is a parsed write request.
+type WriteRequest struct {
+	Database        string
+	RetentionPolicy string
+	// lineProtocol must be valid newline-separated line protocol.
+	lineProtocol []byte
+	// pointOffsets gives the starting index within lineProtocol of each point,
+	// for splitting batches if required.
+	pointOffsets []int
+}
+
+func NewWriteRequest(r *coordinator.WritePointsRequest, log *zap.Logger) (wr WriteRequest, numInvalid int64) {
+	log = log.With(zap.String("database", r.Database), zap.String("retention_policy", r.RetentionPolicy))
+	// Pre-allocate at least smallPointSize bytes per point.
+	const smallPointSize = 10
+	writeReq := WriteRequest{
+		Database:        r.Database,
+		RetentionPolicy: r.RetentionPolicy,
+		pointOffsets:    make([]int, 0, len(r.Points)),
+		lineProtocol:    make([]byte, 0, len(r.Points)*smallPointSize),
+	}
+	numInvalid = 0
+	for _, p := range r.Points {
+		if err := models.ValidPointStrings(p); err != nil {
+			log.Debug("discarding point", zap.Error(err))
+			numInvalid++
+			continue
+		}
+		writeReq.pointOffsets = append(writeReq.pointOffsets, len(writeReq.lineProtocol))
+		writeReq.lineProtocol = p.AppendString(writeReq.lineProtocol)
+		writeReq.lineProtocol = append(writeReq.lineProtocol, byte('\n'))
+	}
+	return writeReq, numInvalid
+}
+
+// pointAt uses pointOffsets to slice the lineProtocol buffer and retrieve the i_th point in the request.
+// It includes the trailing newline.
+func (w *WriteRequest) PointAt(i int) []byte {
+	start := w.pointOffsets[i]
+	end := len(w.lineProtocol)
```

Comment:
Was confused about this until I saw the `if` statement below it. Could you add a short comment above this line calling out the logic?